File: Difficulty: Medium/Substrings with K Distinct/substrings-with-k-distinct.cpp
class Solution {
  public:
    int countAtMostK(string& s, int k) {
        int n           = s.length(); // Length of the string
        int mp[26]      = { 0 };      // Frequency array for lowercase letters
        int i           = 0;          // Left pointer of sliding window
        int j           = 0;          // Right pointer of sliding window
        int result      = 0;          // Count of substrings
        int distinctCnt = 0;          // Number of distinct characters in current window

        while (j < n) {
            // If the character is new to the window
            if (mp[s[j] - 'a'] == 0) {
                distinctCnt++;
            }
            mp[s[j] - 'a']++; // Add current character to the window

            // Shrink the window until we have at most k distinct characters
            while (distinctCnt > k) {
                mp[s[i] - 'a']--;  // Remove from left
                if (mp[s[i] - 'a'] == 0) {
                    distinctCnt--; // A distinct character is completely removed
                }
                i++;
            }

            result += j - i + 1; // All substrings ending at j with at most k distinct
            j++;
        }
        return result;
    }
    int countSubstr(string& s, int k) {
        // Count substrings with exactly k distinct characters
        return countAtMostK(s, k) - countAtMostK(s, k - 1);
    }
};

File: Difficulty: Medium/Substrings with K Distinct/substrings-with-k-distinct.java
// User function Template for Java

class Solution {
    int countAtMostK(String s, int k) {
        int n = s.length();                 // Length of the string
        int[] freq = new int[26];           // Frequency array for 'a' to 'z'
        int i = 0, j = 0;                   // Two pointers
        int result      = 0;                // Count of substrings
        int distinctCnt = 0;                // Number of distinct characters

        while (j < n) {
            if (freq[s.charAt(j) - 'a'] == 0) {
                distinctCnt++;              // New distinct character
            }
            freq[s.charAt(j) - 'a']++;      // Add character to the window

            while (distinctCnt > k) {
                freq[s.charAt(i) - 'a']--;  // Shrink from the left
                if (freq[s.charAt(i) - 'a'] == 0) {
                    distinctCnt--;          // A distinct character is completely removed
                }
                i++;
            }

            result += j - i + 1;            // Valid substrings ending at j
            j++;
        }

        return result;
    }
    int countSubstr(String s, int k) {
        return countAtMostK(s, k) - countAtMostK(s, k - 1); // Exactly K distinct substrings
    }
}